File: 220919/JavaScript/n-queen.js
function isPromising({ x, y, columns }) {
  for (let i = 0; i < x; i += 1) {
    if (columns[i] === y || Math.abs(x - i) === Math.abs(y - columns[i])) {
      return false;
    }
  }

  return true;
}

function depthFirstSearch({
  length, columns, row, answer,
}) {
  if (length === row) {
    answer += 1;
    return answer;
  }

  for (let i = 0; i < length; i += 1) {
    columns[row] = i;

    if (isPromising({ x: row, y: i, columns })) {
      answer = depthFirstSearch({
        length,
        columns,
        row: row + 1,
        answer,
      });
    }
  }

  return answer;
}

export default function solution(length) {
  let answer = 0;

  const columns = Array(length).fill(0);

  answer = depthFirstSearch({
    length,
    columns,
    row: 0,
    answer,
  });

  return answer;
}

File: solutions/Previous/220919/Java/NQueen.java
public class NQueen {
  int answer = 0;

  public int solution(int length) {
    int[] placedPositionX = new int[length];
    int[] placedPositionY = new int[length];

    for (int i = 0; i < length; i += 1) {
      placedPositionX[0] = 0;
      placedPositionY[0] = i;

      putQueen(placedPositionX, placedPositionY, 1, length);
    }

    return answer;
  }

  public void putQueen(int[] placedPositionX,
                       int[] placedPositionY,
                       int count, int length) {
    if (count == length) {
      answer += 1;
      return;
    }

    for (int i = 0; i < length; i += 1) {
      int newPositionX = count;
      int newPositionY = i;
      boolean isPossibleCase = true;

      for (int j = 0; j < count; j += 1) {
        int previousPositionX = placedPositionX[j];
        int previousPositionY = placedPositionY[j];

        if (previousPositionY == newPositionY) {
          isPossibleCase = false;
          break;
        }

        if (Math.abs(newPositionX - previousPositionX)
            == Math.abs(newPositionY - previousPositionY)) {
          isPossibleCase = false;
          break;
        }
      }

      if (isPossibleCase) {
        placedPositionX[count] = newPositionX;
        placedPositionY[count] = newPositionY;

        putQueen(placedPositionX, placedPositionY, count + 1, length);
      }
    }
  }
}
